Make `Numeric` report compliant

konsumlamm opened the pull request:
Make `Numeric` report compliant by exporting more functions that were already there and newly defining `readFloat` and `fromRat`. The `readFloat` code is taken verbatim from GHC base, the `fromRat` implementation is adapted from the GHC version. `Numeric` now imports `Prelude`, since I didn't want to write down all the needed imports, I hope that's okay .

augustss commented on lib/Numeric.hs:
```diff
@@ -1,14 +1,113 @@
+wordLog2 w = _wordSize - 1 - primWordClz w
+
+integerLog2 :: Integer -> Int
+integerLog2 (I sign ds) =
```
This won't work if you compile to use GMP.  The internal representation of Integer should never be used.
Perhaps the integerLog stuff should move to some other module.

I also think that the fromRat code is much more complex than necessary (I don't mind saying that, since I'm the original author of the code, almost 35 years ago).  I think a variation of `fromRational` from `Fractional Double` would be better.

konsumlamm replied:
> This won't work if you compile to use GMP.

Right, of course. There should probably be a CI job using GMP.

> I think a variation of `fromRational` from `Fractional Double` would be better.

Then I have some questions on [that code](https://github.com/augustss/MicroHs/blob/e72d01f5be926255fe25320f2fec83bdc90ce919/lib/Data/Double.hs#L49-L64):
- Why 1023 instead of 1024 (which is the maximum exponent)?
- Why require `d < maxVal` instead of `d <= maxVal`?
- How was 256 chosen as shift?

I don't think anyone really uses `fromRat` (most people probably don't even know that it exists), so it's not important for it to be very efficient. I'm happy to go with your suggestion, if that's what you prefer.

augustss replied:
Yes, a CI job for GMP would be great.  It really only has to be run on one platform.

Answers: all the answers are that it was just accidental.  It doesn't really matter exactly what the constants are, as long as they scale numbers that are too big.  It should be possible to figure out some better set of constants from the `RealFloat` information.  In which case `Double` and `Float` could share a `fromRational` implementation.

konsumlamm commented:
I removed the `integerLog2` special case and added a shortcut, if the numerator and denominator fit in a `RealFloat a => a`. I'll see if an implementation based on `fromRational @Double` works better.
